src/Interactive/EditorFeatures/Core/Extensibility/Interactive/SendToInteractiveSubmissionProvider.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Editor.OptionsExtensionMethods;
+
+namespace Microsoft.CodeAnalysis.Editor.Interactive
+{
+    /// <summary>
+    /// Implementers of this interface are responsible for retrieving source code that
+    /// should be sent to the REPL given the user's selection.
+    /// 
+    /// If the user does not make a selection then a line should be selected.
+    /// If the user selects code that fails to be parsed then the selection gets expanded
+    /// to a syntax node.
+    /// </summary>
+    internal abstract class SendToInteractiveSubmissionProvider : ISendToInteractiveSubmissionProvider
+    {
+        /// <summary>Expands the selection span of an invalid selection to a span that should be sent to REPL.</summary>
+        protected abstract IEnumerable<TextSpan> GetExecutableSyntaxTreeNodeSelection(TextSpan selectedSpan, SourceText source, SyntaxNode node, SemanticModel model);
+
+        /// <summary>Returns whether the submission can be parsed in interactive.</summary>
+        protected abstract bool CanParseSubmission(string code);
+
+        public string GetSelectedText(IEditorOptions editorOptions, CommandArgs args, CancellationToken cancellationToken)
```

Comment:
If no one is expected to call this directly, my personal preference is to use explicit interface implementation to make the relationship clearer.